src/neo/SmartContract/Native/NameService.cs

```diff
@@ -0,0 +1,280 @@
+        public bool IsAvailable(StoreView snapshot, string name)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            string[] names = name.Split('.');
+            if (names.Length != 2) throw new ArgumentException(null, nameof(name));
+            byte[] hash = GetKey(Utility.StrictUTF8.GetBytes(name));
+            if (snapshot.Storages.TryGet(CreateStorageKey(Prefix_Token).Add(hash)) is not null) return false;
+            StringList roots = snapshot.Storages.TryGet(CreateStorageKey(Prefix_Roots))?.GetInteroperable<StringList>();
+            if (roots is null || roots.BinarySearch(names[1]) < 0) throw new InvalidOperationException();
+            return true;
+        }
+
+        [ContractMethod(0_01000000, CallFlags.WriteStates)]
+        private bool Register(ApplicationEngine engine, string name, UInt160 owner)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            string[] names = name.Split('.');
+            if (names.Length != 2) throw new ArgumentException(null, nameof(name));
+            if (!engine.CheckWitnessInternal(owner)) throw new InvalidOperationException();
+            byte[] hash = GetKey(Utility.StrictUTF8.GetBytes(name));
+            if (engine.Snapshot.Storages.TryGet(CreateStorageKey(Prefix_Token).Add(hash)) is not null) return false;
+            StringList roots = engine.Snapshot.Storages.TryGet(CreateStorageKey(Prefix_Roots))?.GetInteroperable<StringList>();
+            if (roots is null || roots.BinarySearch(names[1]) < 0) throw new InvalidOperationException();
+            engine.AddGas(GetPrice(engine.Snapshot));
+            NameState state = new NameState
+            {
+                Owner = owner,
+                Admin = owner,
+                Name = name,
+                Description = "",
+                Expiration = (uint)(engine.Snapshot.PersistingBlock.Timestamp / 1000) + OneYear
+            };
+            Mint(engine, state);
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_Expiration).AddBigEndian(state.Expiration).Add(hash), new StorageItem(new byte[] { 0 }));
+            return true;
+        }
+
+        [ContractMethod(0, CallFlags.WriteStates)]
+        private uint Renew(ApplicationEngine engine, string name)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            string[] names = name.Split('.');
+            if (names.Length != 2) throw new ArgumentException(null, nameof(name));
+            engine.AddGas(GetPrice(engine.Snapshot));
+            byte[] hash = GetKey(Utility.StrictUTF8.GetBytes(name));
+            NameState state = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Token).Add(hash)).GetInteroperable<NameState>();
+            if (!engine.CheckWitnessInternal(state.Owner) || !engine.CheckWitnessInternal(state.Admin)) throw new InvalidOperationException();
```

Comment:
@erikzhang note that I changed it to be OR instead of AND, I think that it's better, what do you think?

Reply:
Then why do we need admin...

Reply:
I think it is enough if any one of them agrees. So it should be `&&`.

Reply:
But with AND, we need to sign all with two accounts

Reply:
Reverted